### PR TITLE
provider/core: include tag metadata, allow partial upsert/deletes

### PR DESCRIPTION
**Pull request by anotherchrisberry**
* allow delete operation to specific a list of keys to delete, or `deleteAll` to remove a tag altogether
* allow partial update of entity tags, e.g. upsert operations against only specified tag members

@ajordens PTAL

### CODE REVIEW

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java [hunk 1]
```diff
@@ -93,6 +83,83 @@ public Void operate(List priorOutputs) {
     return null;
   }
 
+  @VisibleForTesting
+  public UpsertEntityTagsDescription setIdAndIdPattern(UpsertEntityTagsDescription description) {
```
Comment (ajordens) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: If it were me, I'd probably just leave this private and static (_if possible_). 

The groovy test will be able to access it no problem.

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java [hunk 2]
```diff
@@ -51,28 +54,15 @@ public UpsertEntityTagsAtomicOperation(Front50Service front50Service,
...
-    String entityRefAccountId = (String) entityRef.attributes().get("accountId");
-    if (entityRefAccount != null && entityRefAccountId == null) {
-      // add `accountId` if available (and not already specified)
-      AccountCredentials accountCredentials = accountCredentialsProvider.getCredentials(entityRefAccount);
-      if (accountCredentials != null) {
-        entityRefAccountId = accountCredentials.getAccountId();
-        entityRef.attributes().put("accountId", entityRefAccountId);
-      }
+    if (entityTagsDescription.getId() == null) {
+      setIdAndIdPattern(entityTagsDescription);
```
Comment (ajordens) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: `entityTagsDescription = setIdAndIdPattern(entityTagsDescription)`

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java [hunk 3]
```diff
@@ -93,6 +83,83 @@ public Void operate(List priorOutputs) {
...
+      Optional.ofNullable(entityRefAccountId).orElse(entityRefAccount),
+      (String) entityRef.attributes().get("region")
+    );
+    description.setId(entityRefId.id);
+    description.setIdPattern(entityRefId.idPattern);
+    return entityTagsDescription;
+  }
+
+  @VisibleForTesting
+  public Void mergeExistingTagsAndMetadata(EntityTags currentTags, EntityTags updatedTags) {
```
Comment (ajordens) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: `public static EntityTags mergeExistingTagsAndMetadata(EntityTags currentTags, EntityTags updatedTags) {`

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java [hunk 4]
```diff
@@ -93,6 +83,83 @@ public Void operate(List priorOutputs) {
...
+        if (!updatedTags.getTags().containsKey(key)) {
+          updatedTags.getTags().put(key, tag);
+          updatedTags.getTagsMetadata().put(key, currentTags.getTagsMetadata().get(key));
+        }
+      });
+    }
+    return null;
+  }
+
+  private Void addMetadataForTag(EntityTags entityTags, String key) {
```
Comment (ajordens) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: `private static EntityTags.EntityTagMetadata tagMetadata()`

and then on the caller side ...

`entityTags.getTagsMetadata().put(key, tagMetadata())`

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/DeleteEntityTagsAtomicOperation.java [hunk 5]
```diff
@@ -44,13 +45,40 @@ public DeleteEntityTagsAtomicOperation(Front50Service front50Service,
 
   @Override
   public Void operate(List priorOutputs) {
-    getTask().updateStatus(BASE_PHASE, format("Deleting %s from ElasticSearch", entityTagsDescription.getId()));
-    entityTagsProvider.delete(entityTagsDescription.getId());
-    getTask().updateStatus(BASE_PHASE, format("Deleted %s from ElasticSearch", entityTagsDescription.getId()));
+    getTask().updateStatus(BASE_PHASE, format("Retrieving %s from Front50", entityTagsDescription.getId()));
+    EntityTags currentTags = front50Service.getEntityTags(entityTagsDescription.getId());
```
Comment (ajordens) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/DeleteEntityTagsAtomicOperation.java]: front50Service.getEntityTags() would actually throw an exception vs. returning null (standard retrofit behavior)

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java (hunk 2 above)
Comment (anotherchrisberry) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: `entityTagsDescription` is marked as `final` - does it make sense to change that?

I don't think removing `final` will cause any underlying issues, but assigning the return value is a little misleading, as it implies (to me, at least) that the `setIdAndIdPattern` method returns a _different_ object than the one passed in, when it just performs some mutation.

I feel like I'm missing some pattern/practical context here.

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java (hunk 4 above)
Comment (anotherchrisberry) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/UpsertEntityTagsAtomicOperation.java]: 👍

File: clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/DeleteEntityTagsAtomicOperation.java (hunk 5 above)
Comment (anotherchrisberry) [inline clouddriver-elasticsearch/src/main/java/com/netflix/spinnaker/clouddriver/elasticsearch/ops/DeleteEntityTagsAtomicOperation.java]: yep, good point, thanks for catching

**Pull request merged**